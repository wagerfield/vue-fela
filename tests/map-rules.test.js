import isArray from 'isarray'
import isObject from 'isobject'
import { createRenderer } from 'fela'
import { mapRules } from '../index'
import {
  testSnapshot,
  installPlugin,
  wrapComponent
} from './test-helpers'
import props from './component-props'
import rules from './component-rules'

const renderer = createRenderer()
const localVue = installPlugin({ renderer })

const props1 = { outerRadius: 0, innerRadius: 0, active: false }
const props2 = { outerRadius: 4, innerRadius: 2, active: true }

const createComponent = (optMap, optProps) => ({
  props,
  render(createElement) {
    const keys = isArray(optMap)
      ? optMap
      : isObject(optMap)
        ? Object.keys(optMap)
        : Object.keys(rules)
    return createElement('div', {
      class: keys.map((key) => this[key])
    })
  },
  computed: mapRules(rules, optMap, optProps)
})

const testShape = (optMap, shape) => {
  it('returns expected shape', () => {
    expect(mapRules(rules, optMap)).toEqual(shape)
  })
}

const testPropsData = (optMap, propsData) => {
  it('renders expected snapshots from propsData', () => {
    const component = createComponent(optMap)
    const wrapper = wrapComponent(component, localVue, { propsData })
    testSnapshot(wrapper)
  })
}

const testOptProps = (optMap, optProps) => {
  it('renders expected snapshots from optProps', () => {
    const component = createComponent(optMap, optProps)
    const wrapper = wrapComponent(component, localVue)
    testSnapshot(wrapper)
  })
}

const testPrecedence = (optMap, optProps, propsData) => {
  it('optProps takes precedence over propsData', () => {
    const component = createComponent(optMap, optProps)
    const wrapper1 = wrapComponent(component, localVue)
    const wrapper2 = wrapComponent(component, localVue, { propsData })
    expect(wrapper1.vm.className).toBe(wrapper2.vm.className)
  })
}

const testSuite = (shape, optMap) => {
  describe(`optMap = ${JSON.stringify(optMap)}`, () => {
    testShape(optMap, shape)
    testOptProps(optMap, props1)
    testPropsData(optMap, props1)
    testPrecedence(optMap, props1, props2)
  })
}

describe('mapRules(rule, optMap, optProps)', () => {

  beforeEach(renderer.clear)

  it('returns an object', () => {
    expect(mapRules(rules)).toEqual(expect.any(Object))
  })

  testSuite({
    border: expect.any(Function),
    layout: expect.any(Function)
  })

  testSuite({
    border: expect.any(Function)
  }, [
    'border'
  ])

  testSuite({
    borderAlias: expect.any(Function)
  }, {
    borderAlias: 'border'
  })
})
