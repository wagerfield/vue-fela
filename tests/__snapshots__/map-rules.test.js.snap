// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapRules(rule, optMap, optProps) optMap = ["border"] renders expected snapshots from optProps 1`] = `<div class="a b c"></div>`;

exports[`mapRules(rule, optMap, optProps) optMap = ["border"] renders expected snapshots from propsData 1`] = `<div class="a b c"></div>`;

exports[`mapRules(rule, optMap, optProps) optMap = {"borderAlias":"border"} renders expected snapshots from optProps 1`] = `<div class="a b c"></div>`;

exports[`mapRules(rule, optMap, optProps) optMap = {"borderAlias":"border"} renders expected snapshots from propsData 1`] = `<div class="a b c"></div>`;

exports[`mapRules(rule, optMap, optProps) optMap = undefined renders expected snapshots from optProps 1`] = `<div class="a b c d e f"></div>`;

exports[`mapRules(rule, optMap, optProps) optMap = undefined renders expected snapshots from propsData 1`] = `<div class="a b c d e f"></div>`;
